pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -1419,8 +1419,15 @@ public LLRealtimeSegmentDataManager(SegmentZKMetadata segmentZKMetadata, TableCo
 
     // Create message decoder
     Set<String> fieldsToRead = IngestionUtils.getFieldsForRecordExtractor(_tableConfig.getIngestionConfig(), _schema);
-    StreamMessageDecoder streamMessageDecoder = StreamDecoderProvider.create(_partitionLevelStreamConfig, fieldsToRead);
-    _streamDataDecoder = new StreamDataDecoderImpl(streamMessageDecoder);
+    try {
+      StreamMessageDecoder streamMessageDecoder =
+          StreamDecoderProvider.create(_partitionLevelStreamConfig, fieldsToRead);
+      _streamDataDecoder = new StreamDataDecoderImpl(streamMessageDecoder);
+    } catch (Exception e) {
+      _realtimeTableDataManager.addSegmentError(_segmentNameStr,
+          new SegmentErrorInfo(now(), "Failed to initialize the StreamMessageDecoder", e));
```

Comment:
why do we log to the table data manager here instead of using `_segmentLogger.error` API here?

or the other way to ask is: there are other _segmentLogger.error or thrown error in this method, don't we also need to call `_realtimeTableDataManager.addSegmentError` to those as well?

if the comment above make sense, should we do something similar to `PartitionConsumer` consume loop and call the _realtimeTableDataManager.addSegmentError at the very end as a catch all?

Reply:
Yeah I know what you mean. This class requires a lot of rework.

> there are other _segmentLogger.error or thrown error in this method, don't we also need to call _realtimeTableDataManager.addSegmentError to those as well?

Yeah. I think I can add one more segment error in the last try-catch block in the constructor.  

 > if the comment above make sense, should we do something similar to PartitionConsumer consume loop and call the _realtimeTableDataManager.addSegmentError at the very end as a catch all?

The partition consumer thread already adds to the segment error cache.

Reply:
>yes, what i meant is doing the same thing in this method as how PartitionConsumer is doing in its run method

Given the complexity of this constructor, I wouldn't bucket everything into a single catch block. I think it also helps provide more accurate error message , even before looking at the stacktrace.